CIP-???? | Authenticated Web3 HTTP requests (#442)
cardano-foundation/CIPs: CIP-????/README.md

---
CIP: ?
Title: Authenticated Web3 HTTP requests
Category: Tools
Status: Proposed
Authors:
    - Juan Salvador Magán Valero <jmaganvalero@gmail.com>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/cips/pulls/?
Created: 2022-12-27
License: CC-BY-4.0
---

# CIP-XXXX: Authenticated Web3 HTTP requests

## Abstract

The cardano wallets have the ability to sign arbitrary pieces of data as we can see in the [Message signing CIP-0008](/CIP-0008/README.md). All wallets implement the method ```api.signData(addr: Address, payload: Bytes): Promise<DataSignature>``` defined in [Cardano dApp-Wallet Web Bridge CIP-0030](/CIP-0030/README.md).

> An important use case for web3 sites is calling HTTP endpoints using authenticated requests signed by the private keys of the users' wallets. In this way, the servers can perform actions using onchain and offchain data. 

Author: When I say the server, I mean the own dApp application server. I pursuit real decentralization, only the App server and the Blockchain is needed for this Authentication model. It can be also  used for authorization using NFTs.

I know that lately almost all Web3 Blockchain authorization and authentication models use external entities as identity authorities. But this CIP isn't about it.

I want to provide a simple and effective method for authentication without additional actors.

> An important use case for web3 sites is calling HTTP endpoints using authenticated requests signed by the private keys of the users' wallets. In this way, the servers can perform actions using onchain and offchain data. 

Author: @KtorZ I rewrote the abstract. Sorry for the delay. Could you please check it?

> An important use case for web3 sites is calling HTTP endpoints using authenticated requests signed by the private keys of the users' wallets. In this way, the servers can perform actions using onchain and offchain data. 

Reviewer: Disambiguated as per https://github.com/cardano-foundation/CIPs/pull/442#discussion_r1182508929 and @jmagan I can see how the Abstract now includes some functional description.  It's still a bit vague and redundant, so could be improved... will leave it to @KtorZ about whether it's currently acceptable.  I would like to see one clear sentence about what the proposed component is actually doing, with less general information about Cardano and CIPs.

> An important use case for web3 sites is calling HTTP endpoints using authenticated requests signed by the private keys of the users' wallets. In this way, the servers can perform actions using onchain and offchain data. 

Author: I was thinking about how to improve the Abstract part but I'm clueless. I can't figure out how to write it without referencing Cardano. Shouldn't it be general and easy to understand?

> An important use case for web3 sites is calling HTTP endpoints using authenticated requests signed by the private keys of the users' wallets. In this way, the servers can perform actions using onchain and offchain data. 

Reviewer: no @jmagan by all means I wasn't suggesting you not mention Cardano: just that it doesn't have to read like an article for an open audience.  If you want to set a more conversational tone, that's up to you.

## Motivation

dApps generate arbritary payloads as byte arrays. These payloads are signed and included in the protected headers of the signatures. The wallets are responsible for showing the payload data to the user, who will proceed to sign or reject the payload. It's a common practice to encode a string or a JSON string but there isn't any standard for the way to construct and to show this data.

The current implementations for web3 applications use static strings. This is dangerous because if a bad actor intercepts the signed message then it can be used in a replay attack by the bad actor. That's why it is very important to produce a dynamic payload rather a static string.

Another problem with the current approach is how the wallets show the information contained in the payload. The payload is a enconded byte array and it could contain anything. If Alice want to call an endpoint and Bob has the ability to change the message before Alice gets it. Alice must be notified somehow that she is signing a potentially malicious payload. A simple hex-encoded representation of the payload isn't enough to ensure a safe interaction.

## Specification

> This specification involves multiple parties. We split this specification in three parts: server processing, payload structure and wallet specification.

Reviewer 2: It isn't clear to me what "the server" refers to in this proposal. I think the proposal would deserve a bit of an introduction regarding each actor involved and their role.

Author: I added dApp Server and the ASCII diagram. I believe now it's clearer.

Author: @rphair I added some definition before the ASCII diagram:

> This specification involves multiple parties: Wallet/Client, dApp Server and Blockchain. 
> 
> 1. **Wallet/Client**: The Wallet/Client is responsible for managing the user's cryptographic keys. Anyone can create a wallet using the CIP-0030 API interface, but it may produce invalid or malicious data sent to the dApp. This CIP aims to validate the ownership and veracity of the data provided by wallets. Additionally, it establishes guidelines for mitigating common wallet attacks, improving security for user interaction.
> 
> 2. **dApp Server**: The dApp Server represents the server-side infraestructure that supports decentralized applications (dApps). It communicates with the blockchain to retrieve stored data and validate wallet status. It must enforce minimum payload requirements to ensure authenticity and protect users from malicious actors.
> 
> 3. **Blockchain**: The Blockchain is the underlying distributed ledger technology that forms the foundation of decentralized systems. It is a decentralized and immutable ledger that securely records all transactions and data in a chronological and transparent manner. The Blockchain can be utilized for authentication, providing user identity, and for authorization, tracking user history and current status.

I believe this gives more background and context to each actor.

### Payload structure

The payload MUST be encoded as a JSON string. JSON strings are semi-structured data that are human-readable. So that with a straightforward decoding into text, it will be understandable for the readers. This feature also allows the users to debug it in an easy manner, for example, with the browser debugger tools.

The content of the payload will be included in the protected header of the COSESign1 signature, hence the content effects directly the behavior and security of the system. The payload MUST have the following fields: 

1. The field ```url``` MUST contain the full path to the endpoint, where the payload will be processed.

2. Sometimes, the endpoint ```url``` field is not enough to determine its purpose. The user should understand perfectly the objective of the payload which he or she is signing. That's why the payload MUST contain an ```action``` field with a descriptive text containing the purpose of the payload. 

> 3. The payload MUST include a UNIX timestamp. The ```timestamp``` field is used as a nonce and a mark for the payload expiration. This field is very imporant in case the payload is compromised.

Reviewer 2: How could the payload be "compromised" here? What is the assumed threat model?

Author: You can call the endpoint, for example "http://my-awesome-dapp/users", but this endpoint can create or delete users. If someone gets the signed payload, he or she can access to both functions. This action field mitigates risk.

Reviewer: @jmagan this is understandable, but still not explained in the CIP as such: can you add a little paragraph repeating or explaining what you've just said?  cc @KtorZ

Author: @KtorZ I added an example and how it's mitigated this threat model as commented by @rphair 

> 2. Sometimes, the endpoint `uri` field is not enough to determine its purpose. The user should understand perfectly the objective of the payload which he or she is signing. That's why the payload MUST contain an `action` field with a descriptive text containing the purpose of the payload. For example, if someone calls the endpoint `/users` without an action field, they can create or delete users. However, by including the action field in the payload, it not only provides additional clarity but also effectively limits the scope of the payload.

Additional fields MAY be included in the payload. Depending on the process, it could be useful to include some aditional fields in the protected header of the signature. For example, the email information should be included in the protected header in a registration request. In that way, this payload can be used only to register that specific email and it can not be tampered. 

```JSON
{
    "url": "http://example.com/signup",
    "action": "Sign up",
    "timestamp": 1673261248,
    ["email": "email@example.com"] <- Optional
}
```

### Wallet specification

> The wallets can improve the overall security implementing the following guidelines. We RECOMMEND to show in a structured way the payload information for shake of clarity. This information should be well understanded by the users before the payload is signed.  

Reviewer 2: ```suggestion
The wallets can improve the overall security implementing the following guidelines. We RECOMMEND to show in a structured way the payload information for sake of clarity. This information should be well understood by the users before the payload is signed.  
```

The ```url``` field provides information about the hostname of the application. This hostname MUST be included in the wallet allow list. If a known domain A tries to sign a payload for an unknown domain B, you will be prompted with permission popup making more obvious the cross-domain interaction. When possible, the wallet SHOULD warn the user if a payload is for a different domain.

The wallet SHOULD update the ```timestamp``` field to the current time just before the signature. This field ideally should match the moment just before the signature, thereby the server recieves the more updated payload possible. 

### Server processing

The server has ultimate responsibility of processing correctly the requests. We use the content to validate the payload. The request will be processed with the following steps:

> 1. The server MUST check the action and the endpoint included in the request. Each route to an endpoint MUST have an associated action and a URL. The first step is to check that they match with the parameterized action.

Reviewer 2: What request are we talking about here and, how does the server "check the action"? What is being sent to the server, the wallet etc..? 

Perhaps a little diagram (even a plain ASCII one) could help better show the interaction between the actors.

Author: @KtorZ I've already created an ASCII diagram. I hope it clarifies the process.

Reviewer: @jmagan I can see you've added at the beginning of this section (https://github.com/jmagan/CIPs/blob/master/CIP-%3F%3F%3F%3F/README.md#specification) so @KtorZ marking this as resolved.

2. The server MUST check the expiration of the payload. The expiration SHOULD be enough to give time to the user to introduce the wallet password but it SHOULD NOT be too long, we RECOMMEND not more than 5 minutes.

3. The server MUST validate the COSESign1 signature and check that the address inside the protected map of the signature corresponds to the public key in the COSEKey. 

Additionally the server COULD extract the payload content and pass it through the server logic.

> ## Rationale: how does this CIP achieve its goals?

Reviewer 2: One question I'd like to see answered in the 'Rationale' is what value does such an authorization has from a service provider standpoint (assumed to be 'the server' here)? Said differently, why would my-awesome-dapp care that a user authorized something with its wallet? Is the service provider assumed to have some knowledge of the key involved in the authorization and is only expecting a proof of ownership? How is such a proof better than a transaction witness proving the ability to spend associated UTxOs?

Author: @KtorZ, I made some improvement in the Rationale. I think now it's a little bit more specific.

Reviewer: @jmagan what's currently in the Rationale is more specific, with respect to currently available alternatives that are trying to do the same thing, but it doesn't go into as much detail as @KtorZ is requesting above.  If it answers the question about "How does the dApp trust the wallet in this method?" then can you please explain it here and/or in the CIP?  (@KtorZ please mark Resolved if you think it is)

Author: @rphair  Okey I see the point. I though that this point would clarify it: 

>  Moreover, the signature method proposed in this CIP does not require users to spend funds in a transaction, which further lowers the cost and barriers to entry for users.

In relation to your question 

>How does the dApp trust the wallet in this method?

I can add a brief summary of or a referenfe to the CIP-0008 but I though It could be a little bit redundant.

Reviewer: I also thought that sentence was relevant to the question: so maybe a bit more detail, maybe even a subheading on that subject, would make the advantage more apparent to service providers... since generally they're the ones who will be implementing your proposal 🤓

Author: I made a connection between CIP-0008 and this CIP at the beggining of the Rationale. I feel that previously this part was missing and now all make more sense.

> ## Rationale: how does this CIP achieve its goals?
> 
> CIP-0008 enhances authentication by enabling individuals to prove ownership of addresses, identities, and off-chain data through message signing. It provides a reliable means of authentication by allowing individuals to attach a public key to their data and sign messages associated with that data, thereby establishing ownership and ensuring the integrity of the authentication process.
> 
> Additionally, This specification provides the general guidelines and necessary recommendations for performing secure authenticated web3 requests in the Cardano ecosystem. It covers the two main desired characteristics for a secure payload: It must expire and it must be non-static. Moreover, the signature method proposed in this CIP does not require users to spend funds in a transaction, which further lowers the cost and barriers to entry for users.

1. The field `uri` MUST contain the full path to the endpoint, where the payload will be processed.

2. Sometimes, the endpoint `uri` field is not enough to determine its purpose. The user should understand perfectly the objective of the payload which he or she is signing. That's why the payload MUST contain an `action` field with a descriptive text containing the purpose of the payload. 

3. The payload MUST include a UNIX timestamp. The `timestamp` field is used as a nonce and a mark for the payload expiration. This field is very important in case the payload is compromised.

Additional fields MAY be included in the payload. Depending on the process, it could be useful to include some aditional fields in the protected header of the signature. For example, the email information should be included in the protected header in a registration request. In that way, this payload can be used only to register that specific email and it can not be tampered. 

```JSON
{
    "uri": "http://example.com/signup",
    "action": "Sign up",
    "timestamp": 1673261248,

>     ["email": "email@example.com"] <- Optional

Author: @matiwinnetou Thank you for your suggestion! Providing the JSON schema instead of an invalid JSON is indeed a great idea.

>     ["email": "email@example.com"] <- Optional

Author: @matiwinnetou, great idea. It's already implmented:

> #### JSON Schema
> ```JSON
> {
>   "$schema": "http://json-schema.org/draft-07/schema#",
>   "type": "object",
>   "properties": {
>     "uri": {
>       "type": "string",
>       "format": "uri"
>     },
>     "action": {
>       "type": "string"
>     },
>     "timestamp": {
>       "type": "integer"
>     }
>   },
>   "required": ["uri", "action", "timestamp"],
>   "additionalProperties": {
>     "type": "string"
>   }
> }
> ```
> #### Minimum payload:
> 
> ```JSON
> {
>     "uri": "http://example.com/signin",
>     "action": "Sign in",
>     "timestamp": 1673261248,
> }
> ```
> #### Sign up payload example:
> ```JSON
> {
>     "uri": "http://example.com/signup",
>     "action": "Sign up",
>     "timestamp": 1673261248,
>     "email": "email@example.com"
> }
> 
> ```

* [jmagan/cardano-nextjs-web3-skeleton](https://github.com/jmagan/cardano-nextjs-web3-skeleton)
## Path to Active

### Acceptance Criteria

- [X] At least one library should implement this authentication method.
- [ ] The 80% users should have wallets implementing the following requirements:
    1. It MUST detect when the payload is formatted using this specification.
    2. The information contained in the payload MUST be parsed and formatted in the signing pop-up.
    3. The wallet SHOULD update the timestamp just before the payload is signed.
    4. The wallet MUST detect if the URI is in the allow list.
    5. The wallet SHOULD warn the user against cross-domain requests.
- [ ] A detailed documentation about web3 standards should be published. This documentation will include this standard and further best practices for web3 technologies.

### Implementation Plan

- [X] Create a library for processing payload according to this specification.
- [ ] Open a conversation about this specification and its possible improvements.
- [ ] Talk about further web3 standards and new specifications. 
- [ ] Write the documentation for web3 developers.

## Copyright

> This CIP is licensed under [CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode

Reviewer 3: ```suggestion
This CIP is licensed under [CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/legalcode).
```
Small formatting clean up

Additional fields MAY be included in the payload. Depending on the process, it could be useful to include some aditional fields in the protected header of the signature. For example, the email information should be included in the protected header in a registration request. In that way, this payload can be used only to register that specific email and it can not be tampered. 

#### JSON Schema
```JSON
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "uri": {
      "type": "string",
      "format": "uri"
    },
    "action": {
      "type": "string"
    },
    "timestamp": {
      "type": "integer"
    }
  },
  "required": ["uri", "action", "timestamp"],
  "additionalProperties": {
    "type": "string"
  }
}
```
#### Minimum payload:

```JSON
{
    "uri": "http://example.com/signin",
    "action": "Sign in",
    "timestamp": 1673261248,
}
```
#### Sign up payload example:
```JSON
{
    "uri": "http://example.com/signup",
    "action": "Sign up",
    "timestamp": 1673261248,
    "email": "email@example.com"
}

```

> 

Reviewer 4: I have a concern with this approach.
It's using the pre-existing `api.signData` method, and these fields are pretty generic and not very identifying.  
Below it is defined for the wallet to unambiguously determine `It MUST detect when the payload is formatted using this specification.`.
How does the wallet do this, and how does the wallet/dApp guard against false positive/negative detection?

Eg, What should the wallet do if it sees `"uri"` and `"action"` but not `"timestamp"`?
1. Assume it is an Authenticated WEB3 HTTP Request that is in error? ; OR
2. Assume that it is NOT an Authenticated WEB3 HTTP Request?

And what should the wallet do in either case?

Author: Hi @stevenj! Apologies for the delay. This functionality seamlessly integrates with the existing implementation. To detect the payload, we simply retrieve a JSON object and check for the presence of the mandatory fields: "uri", "action", and "timestamp," as you mentioned earlier. It's important to note that it should not throw an error because CIP-0030 allows signing arbitrary payloads. The crucial aspect here is that the wallet effectively communicates to the user the specific endpoint and corresponding action associated with the request.

Reviewer: @stevenj this looks resolved based on last comment but feel free to re-open (we've moving this forward & would like to know if any unresolved issues)

### Version history

| Version | Date      | Author                         | Rationale              |   
|:-------:|-----------|--------------------------------|------------------------|
| v1      |2022-12-27 | Juan Salvador Magán Valero     | Initial release        |

### Reference implementation

* [jmagan/passport-cardano-web3](https://github.com/jmagan/passport-cardano-web3)
* [jmagan/cardano-express-web3-skeleton](https://github.com/jmagan/cardano-express-web3-skeleton)
* [jmagan/cardano-nextjs-web3-skeleton](https://github.com/jmagan/cardano-nextjs-web3-skeleton)
## Path to Active

### Acceptance Criteria

- [X] At least one library should implement this authentication method.
- [ ] The 80% users should have wallets implementing the following requirements:
    1. It MUST detect when the payload is formatted using this specification.
    2. The information contained in the payload MUST be parsed and formatted in the signing pop-up.
    3. The wallet SHOULD update the timestamp just before the payload is signed.
    4. The wallet MUST detect if the URI is in the allow list.

>     5. The wallet SHOULD warn the user against cross-domain requests.

Reviewer 4: These act as an update to CIP-30.  A wallet that does this is not following the specification defined for CIP-30 for signData which does not include these rules for payload data processing.  There would be no way for a dApp to know if a wallet obeyed CIP-30 or CIP-93 because CIP-93 is not currently defined as an extension of CIP-30.

Author: I partially addressed this in the previous comment. The proposed changes do not modify the API, and the signData method retains the same arguments and return value. This CIP involves three key parties: the wallet, the payload, and the server processing. It would be challenging to cover all these aspects within a wallet-specific CIP alone, which is why it is presented as a comprehensive specification covering the entire process.

Reviewer: likewise resolved as per https://github.com/cardano-foundation/CIPs/pull/442#discussion_r1235142130.

---
CIP: ?
Title: Authenticated Web3 HTTP requests
Category: Tools
Status: Proposed
Authors:
    - Juan Salvador Magán Valero <jmaganvalero@gmail.com>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/cips/pulls/?
Created: 2022-12-27
License: CC-BY-4.0
---

# CIP-XXXX: Authenticated Web3 HTTP requests

## Abstract

> The cardano wallets have the ability to sign arbitrary pieces of data as we can see in the [Message signing CIP-0008](/CIP-0008/README.md). All wallets implement the method ```api.signData(addr: Address, payload: Bytes): Promise<DataSignature>``` defined in [Cardano dApp-Wallet Web Bridge CIP-0030](/CIP-0030/README.md).

Reviewer 5: Small detail: it is actually CIP-8 combined with CIP-30.

1. The field ```url``` MUST contain the full path to the endpoint, where the payload will be processed.

2. Sometimes, the endpoint ```url``` field is not enough to determine its purpose. The user should understand perfectly the objective of the payload which he or she is signing. That's why the payload MUST contain an ```action``` field with a descriptive text containing the purpose of the payload. 

3. The payload MUST include a UNIX timestamp. The ```timestamp``` field is used as a nonce and a mark for the payload expiration. This field is very imporant in case the payload is compromised.

Additional fields MAY be included in the payload. Depending on the process, it could be useful to include some aditional fields in the protected header of the signature. For example, the email information should be included in the protected header in a registration request. In that way, this payload can be used only to register that specific email and it can not be tampered. 

```JSON
{
    "url": "http://example.com/signup",
    "action": "Sign up",
    "timestamp": 1673261248,

>     ["email": "email@example.com"] <- Optional

Reviewer 5: From GDPR (europe wise) point of view even putting email here as optional is a total disaster...

Author: Maybe it's a bad example. I don't know much about the GDPR. Could you please tell me more about why including the email in the payload can be a problem for the GDPR?

Reviewer: Once the email address is in there, it's personally identifiable data that cannot be removed.  According to my understanding, even if an email address is added optionally it also has to be removable later.

- Technical recommendations (with lots of references): https://www.nature.com/articles/s41598-022-19341-y
- Bureaucratic view from EU itself & recommendations: https://www.europarl.europa.eu/RegData/etudes/STUD/2019/634445/EPRS_STU(2019)634445_EN.pdf

Reviewer: Once the email address is in there, it's personally identifiable data that cannot be removed.  According to my understanding, even if an email address is added optionally it also has to be removable later: even if you have to destroy the whole blockchain to efface it.

- Technical recommendations (with lots of references): https://www.nature.com/articles/s41598-022-19341-y
- Bureaucratic view from EU itself & recommendations: https://www.europarl.europa.eu/RegData/etudes/STUD/2019/634445/EPRS_STU(2019)634445_EN.pdf

Author: Thanks for the reply. Now I see your point but the payload is signed in the client and sent to the server. It isn't stored in the blockchain. So, the GDPR doesn't apply in this case. Right?

Reviewer: > @jmagan: It isn't stored in the blockchain
As far as I'm concerned that takes care of it.  Any other situations where personal data might be exposed off-chain can be dealt with later if they're a problem. 😎

Author: Great. Then can I mark as resolve?

Reviewer: Yes (I'm marking it) 😎

}
```

### Wallet specification

> The wallets can improve the overall security implementing the folliwing guidelines. We RECOMMEND to show in a structured way the payload information for shake of clarity. This information should be well understanded by the users before the payload is singed.  

Reviewer 5: folliwing -> following

Author: Resolve, thanks!

2. The server MUST check the expiration of the payload. The expiration SHOULD be enough to give time to the user to introduce the wallet password but it SHOULD NOT be too long, we RECOMMEND not more than 5 minutes.

3. The server MUST validate the COSESign1 signature and check that the address inside the protected map of the signature corresponds to the public key in the COSEKey. 

Additionally the server COULD extract the payload content and pass it through the server logic.

## Rationale: how does this CIP achieve its goals?

This specification provides the general guidelines and necessary recommendations for performing secure authenticated web3 requests in the Cardano ecosystem. It covers the two main desired characteristics for a secure payload: It must expire and it must be non-static.

Another important aspect for security is how wallets process the payload. They can improve the security using the data inside the payload to warn the users about possible malicious interactions. This specification emphasizes the importance of informing users clearly about the purpose of the payloads and how wallets can use the URL field to apply allow-lists and/or cross-domain policies.

> Finally, it establishes the requirements and recommendations for server side processing. The server must also ensure the validity of the signature and the payload, as well as of its purpose in order to accomplish the authentication. 

Reviewer 5: Would a user have to sign a transaction for every http request or just like JWT there would be a way to cache and store token somewhere and use this until it expires. 

If this is each time then it could be very annoying, usually a webapp makes many requests to the server side, even for one page.

Author: Exactly. This specification isn't for each request. You can use it only for the important requests like Login, Register, etc... Afterwards a JWT token can take control of the session.

I developed an example using this methodology [cardano-express-web3-skeleton)](https://github.com/jmagan/cardano-express-web3-skeleton)

Author: Is good for you if I mark this as resolve?

> ## Reference implementation

Reviewer: ```suggestion
### Reference implementation
```
There isn't a main section heading for this () which could be fixed by demoting it to be part of the Rationale.

Author: I indent it as you propused. Thanks

* [jmagan/passport-cardano-web3](https://github.com/jmagan/passport-cardano-web3)
* [jmagan/cardano-express-web3-skeleton](https://github.com/jmagan/cardano-express-web3-skeleton)

> 

Reviewer: We also need a _Path to Active_ section with acceptance & implementation sub-sections: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0001#statuses

There's a candidate proposal in a similar field with good checklists for these that might be helpful as a model (pending CIP-0045 if this gets merged & branch disappears): https://github.com/fabianbormann/CIPs/blob/master/CIP-%3F%3F%3F%3F/README.md#path-to-active

Author: Thanks for the reference. It was very useful. I wrote the Path to Active.
